src/components/views/avatars/MemberStatusMessageAvatar.js

```diff
@@ -40,77 +40,91 @@ export default class MemberStatusMessageAvatar extends React.Component {
             throw new Error("Cannot use MemberStatusMessageAvatar on anyone but the logged in user");
         }
+        if (!SettingsStore.isFeatureEnabled("feature_custom_status")) {
+            return;
+        }
+        const { user } = this.props.member;
+        if (!user) {
+            return;
+        }
+        user.on("User._unstable_statusMessage", this._onStatusMessageCommitted);
     }
 
-    componentDidMount() {
-        MatrixClientPeg.get().on("RoomState.events", this._onRoomStateEvents);
-
-        if (this.props.member.user) {
-            this.setState({message: this.props.member.user._unstable_statusMessage});
-        } else {
-            this.setState({message: ""});
+    componentWillUmount() {
+        const { user } = this.props.member;
+        if (!user) {
+            return;
         }
+        user.removeListener(
+            "User._unstable_statusMessage",
+            this._onStatusMessageCommitted,
+        );
     }
 
-    componentWillUnmount() {
-        if (MatrixClientPeg.get()) {
-            MatrixClientPeg.get().removeListener("RoomState.events", this._onRoomStateEvents);
+    get hasStatus() {
+        const { user } = this.props.member;
+        if (!user) {
+            return false;
         }
+        return !!user._unstable_statusMessage;
     }
 
-    _onRoomStateEvents = (ev, state) => {
-        if (ev.getStateKey() !== MatrixClientPeg.get().getUserId()) return;
-        if (ev.getType() !== "im.vector.user_status") return;
-        // TODO: We should be relying on `this.props.member.user._unstable_statusMessage`
-        // We don't currently because the js-sdk doesn't emit a specific event for this
-        // change, and we don't want to race it. This should be improved when we rip out
-        // the im.vector.user_status stuff and replace it with a complete solution.
-        this.setState({message: ev.getContent()["status"]});
+    _onStatusMessageCommitted = () => {
+        // The `User` object has observed a status message change.
+        this.setState({
+            hasStatus: this.hasStatus,
+        });
     };
 
     _onClick = (e) => {
         e.stopPropagation();
 
         const elementRect = e.target.getBoundingClientRect();
 
-        // The window X and Y offsets are to adjust position when zoomed in to page
-        const x = (elementRect.left + window.pageXOffset) - (elementRect.width / 2) + 3;
-        const chevronOffset = 12;
-        let y = elementRect.top + (elementRect.height / 2) + window.pageYOffset;
-        y = y - (chevronOffset + 4); // where 4 is 1/4 the height of the chevron
+        const x = (elementRect.left + window.pageXOffset);
+        const chevronWidth = 16; // See .mx_ContextualMenu_chevron_bottom
+        const chevronOffset = (elementRect.width - chevronWidth) / 2;
+        const chevronMargin = 1; // Add some spacing away from target
+        const y = elementRect.top + window.pageYOffset - chevronMargin;
```

Comment:
All these changes look right, but I am particularly cautious about changing the x/y offset: does zooming into the page cause the chevron to drift with the new values?

Reply:
So, it _does_ drift if zoom while the context menu is open, but that happens for me both before and after these changes. In both Firefox and Chrome, `page*Offset` is always 0 for me, so I wasn't quite sure how this was addressing zoom.

How about we open an issue to address the drift on zoom separately?

Reply:
Okay, I filed https://github.com/vector-im/riot-web/issues/8129 for this.